### ISSUE TITLE
`RuntimeError: compile failed` with cuda "__bf16" is undefined (cuda 12.2, ubuntu 22.04)

### ISSUE THREAD

**Issue by barisozmen** [tinygrad/tinygrad#2812]
Hi,

I ran `python3 examples/coder.py` and got a compilation error with cuda. Below are my steps with all the details. I would be glad if you can point out where the issue came from. New to the repo and cannot make sense the error message. 

Steps to reproduce on ubuntu 22.04 machine with cuda 12.2

### 2. Install dependencies
```$ python3 -m pip install -e . && pip3 install sentencepiece```
### 3. run coder.py (and get the error below)
```$ python3 examples/coder.py

create model: 101.23 ms
download weights: 25.95 ms
ram used:  0.00 GB, layers.0.attention.wq.weight                      :   0%| |
loaded weights in 20.52 ms, 0.03 GB loaded at 1.64 GB/s
weights -> model: 67.99 ms
Traceback (most recent call last):
  File "/home/baris/development/fun/tinygrad/examples/coder.py", line 37, in <module>
    nn.state.load_state_dict(model, fix_bf16(convert_from_huggingface(part1, model, 32, 8)), strict=False)
  File "/home/baris/development/fun/tinygrad/tinygrad/nn/state.py", line 68, in load_state_dict
    v.assign(state_dict[k].to(v.device)).realize()
  File "/home/baris/development/fun/tinygrad/tinygrad/tensor.py", line 103, in realize
    run_schedule(self.lazydata.schedule())
  File "/home/baris/development/fun/tinygrad/tinygrad/realize.py", line 28, in run_schedule
    prg = lower_schedule_item(si)
  File "/home/baris/development/fun/tinygrad/tinygrad/realize.py", line 19, in lower_schedule_item
    return Device[si.out.device].get_runner(si.ast)
  File "/home/baris/development/fun/tinygrad/tinygrad/device.py", line 314, in get_runner
    def get_runner(self, ast:LazyOp) -> CompiledASTRunner: return self.to_program(self.get_linearizer(ast))
  File "/home/baris/development/fun/tinygrad/tinygrad/device.py", line 286, in to_program
    return CompiledASTRunner(k.ast, k.name, src, k.global_size, k.local_size, runtime_args).build(self.compiler, self.runtime)
  File "/home/baris/development/fun/tinygrad/tinygrad/device.py", line 254, in build
    self.lib = compiler.__wrapped__(self.prg) if getenv("DISABLE_COMPILER_CACHE") else compiler(self.prg)
  File "/home/baris/development/fun/tinygrad/tinygrad/helpers.py", line 261, in wrapper
    return diskcache_put(table, key, func(*args, **kwargs))
  File "/home/baris/development/fun/tinygrad/tinygrad/runtime/ops_cuda.py", line 23, in compile_cuda
    def compile_cuda(prg) -> bytes: return compile_cuda_style(prg, [f'--gpu-architecture={CUDADevice.default_arch_name}', "-I/usr/local/cuda/include", "-I/usr/include"], cuda.nvrtcProgram, cuda.nvrtcCreateProgram, cuda.nvrtcCompileProgram, cuda.nvrtcGetPTX, cuda.nvrtcGetPTXSize, cuda.nvrtcGetProgramLog, cuda.nvrtcGetProgramLogSize, check)  # noqa: E501
  File "/home/baris/development/fun/tinygrad/tinygrad/helpers.py", line 322, in compile_cuda_style
    if status != 0: raise RuntimeError(f"compile failed: {get_bytes(prog, get_log_size, get_log, check).decode()}")
RuntimeError: compile failed: <null>(3): error: identifier "__bf16" is undefined
  extern "C" __global__ void E_512_64_8_16_4(unsigned int* data0, const __bf16* data1) {
                                                                        ^

<null>(8): error: identifier "__bf16" is undefined
    __bf16 val0 = data1[(gidx1*64)+(lidx3*4)+((lidx2%2)*262144)+((((gidx0*4)+(lidx2/2))%64)*4096)+((gidx0/16)*524288)];
    ^

<null>(9): error: identifier "__bf16" is undefined
    __bf16 val1 = data1[(gidx1*64)+(lidx3*4)+((lidx2%2)*262144)+((((gidx0*4)+(lidx2/2))%64)*4096)+((gidx0/16)*524288)+1];
    ^

<null>(10): error: identifier "__bf16" is undefined
    __bf16 val2 = data1[(gidx1*64)+(lidx3*4)+((lidx2%2)*262144)+((((gidx0*4)+(lidx2/2))%64)*4096)+((gidx0/16)*524288)+2];
    ^

<null>(11): error: identifier "__bf16" is undefined
    __bf16 val3 = data1[(gidx1*64)+(lidx3*4)+((lidx2%2)*262144)+((((gidx0*4)+(lidx2/2))%64)*4096)+((gidx0/16)*524288)+3];
    ^

5 errors detected in the compilation of "<null>".
```

### Info about my system:
```
$ nvidia-smi

Sun Dec 17 05:31:07 2023
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.129.03             Driver Version: 535.129.03   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA GeForce RTX 4090        Off | 00000000:01:00.0  On |                  Off |
|  0%   39C    P8              23W / 450W |    325MiB / 24564MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+

+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A      1161      G   /usr/lib/xorg/Xorg                          118MiB |
|    0   N/A  N/A      1504      G   /usr/bin/gnome-shell                         78MiB |
|    0   N/A  N/A      4575      G   ...irefox/2987/usr/lib/firefox/firefox      109MiB |
+---------------------------------------------------------------------------------------+
```

```
$ uname -a

Linux uname 6.2.0-39-generic #40~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Thu Nov 16 10:53:04 UTC 2 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by barisozmen** [comment]
probably same issue reported here before: https://github.com/tinygrad/tinygrad/issues/2775

**Comment by barisozmen** [comment]
For the record, `$ DEBUG=2 CUDA=1 python3 extra/gemm/simple_matmul.py` runs fine for me. I'm getting this issue with `examples/coder.py`

```
$ python3 extra/gemm/simple_matmul.py

*** CUDA   rand  seed 1702847144 size 16777216        dtype dtypes.float
*** CUDA   rand  seed 1702847145 size 16777216        dtype dtypes.float
*** CUDA       1 r_128_64_8_16_1024_4_4_4              arg   3 mem  0.20 GB tm   4734.62us/     4.73ms (29028.48 GFLOPS,   42.52 GB/s)
*** CUDA       2 r_128_64_8_16_1024_4_4_4              arg   3 mem  0.27 GB tm   4731.90us/     9.47ms (29045.17 GFLOPS,   42.55 GB/s)
*** CUDA       3 r_128_64_8_16_1024_4_4_4              arg   3 mem  0.27 GB tm   4718.91us/    14.19ms (29125.14 GFLOPS,   42.66 GB/s)
*** CUDA       4 r_128_64_8_16_1024_4_4_4              arg   3 mem  0.27 GB tm   4723.68us/    18.91ms (29095.74 GFLOPS,   42.62 GB/s)
*** CUDA       5 r_128_64_8_16_1024_4_4_4              arg   3 mem  0.27 GB tm   4722.56us/    23.63ms (29102.64 GFLOPS,   42.63 GB/s)
*** CUDA       6 r_128_64_8_16_1024_4_4_4              arg   3 mem  0.27 GB tm   4723.58us/    28.36ms (29096.33 GFLOPS,   42.62 GB/s)
*** CUDA       7 r_128_64_8_16_1024_4_4_4              arg   3 mem  0.27 GB tm   4724.70us/    33.08ms (29089.43 GFLOPS,   42.61 GB/s)
*** CUDA       8 r_128_64_8_16_1024_4_4_4              arg   3 mem  0.27 GB tm   4725.60us/    37.81ms (29083.92 GFLOPS,   42.60 GB/s)
*** CUDA       9 r_128_64_8_16_1024_4_4_4              arg   3 mem  0.27 GB tm   4726.78us/    42.53ms (29076.63 GFLOPS,   42.59 GB/s)
*** CUDA      10 r_128_64_8_16_1024_4_4_4              arg   3 mem  0.27 GB tm   4720.61us/    47.25ms (29114.67 GFLOPS,   42.65 GB/s)
*** CPU       11 copy 16777216,     CPU <- CUDA        arg   2 mem  0.27 GB tm   8235.48us/    55.49ms (    0.00 GFLOPS,    8.15 GB/s)
*** CPU       12 copy 16777216,     CPU <- CUDA        arg   2 mem  0.27 GB tm   8810.30us/    64.30ms (    0.00 GFLOPS,    7.62 GB/s)
*** CPU       13 copy 16777216,     CPU <- CUDA        arg   2 mem  0.27 GB tm     10.69ms/    74.99ms (    0.00 GFLOPS,    6.28 GB/s)
avg: 18328.42 GFLOPS    29.53 GB/s           total:    13 kernels  1374.39 GOPS     2.21 GB    74.99 ms
```

**Comment by RomanVolkov** [comment]
I have similar problem with METAL backend for the same example

```
^~~~~~~~~~~~~~~~~~~~~~~~~~
program_source:3:70: note: type '__bf16' cannot be used in buffer pointee type
kernel void E_512_64_8_16_4(device unsigned int* data0, const device __bf16* data1, uint3 gid [[threadgroup_position_in_grid]], uint3 lid [[thread_position_in_threadgroup]]) {
                                                                     ^
}
```

**Comment by g1y5x3** [comment]
should be fixed with by #3036 for the cuda backend
